### DISCUSSION TITLE
Confirmation on changes of the implementation of dll related errors (api-ms-win-core-file-l2-1-0.dll)

### DISCUSSION THREAD

**Discussion by ghost** [pyinstaller/pyinstaller#7429] [PyInstaller]
On a fresh install of Windows 7 without any vcredists installed, I encountered the error a while ago below when trying to run a Pyinstaller 5.1 onefile binary that was originally built on Windows 10.

```
The program can't start because api-ms-win-core-file-l2-1-0.dll is missing from your computer
```

But this seems to no longer be the case with Pyinstaller 5.7. As a sanity check, can someone confirm if something related to the dll files were changed between versions 5.1-5.7? As far as I can tell, the dll in the error above exists in the extracted directory with both Pyinstaller versions.

**Comment by rokm** [comment]
If there was a change, it was accidental and a side-effect of some other change. 

I did a quick test, though, and an application frozen in python 3.9 environment on Windows 10 does not seem to work on (snapshot of clean installation of) Windows 7, regardless of PyInstaller version. The missing DLL in my case is `api-ms-win-core-path-l1-1-0.dll`. The `api-ms-*` are not collected by default (because they are not resolvable to PyInstaller). They are collected in Visual Studio command line prompt environment (because it adds directory with ucrt DLLs to the `PATH`), but that particula DLL is not among the collected ones. Same goes when freezing in Anaconda-based environment, which has *some* `api-ms-*` DLLs resolvable in its environment, but not this particular one.

So as far as I'm concerned, python 3.9 and later (and thus frozen applications created with them) are not supported on Windows 7. Similarly, it should not be surprising if applications frozen on Windows 10 fail to work on Windows 7.

**Reply by ghost** [reply]
Python 3.9 does not support officially support Windows 7, perhaps your error is related to that rather than Pyinstaller? I froze the application with Python 3.8.6 on Windows 10 which is the last Python version to support Windows 7.

**Reply by rokm** [reply]
> Python 3.9 does not support officially support Windows 7, perhaps your error is related to that rather than Pyinstaller? I froze the application with Python 3.8.6 on Windows 10 which is the last Python version to support Windows 7.

In that case, there's something else you aren't telling us that is making the difference.

Because like you noted yourself, python 3.8 is fully compatible with Windows 7. A basic hello world example frozen with python 3.8 on Windows 10 works on Windows 7 regardless of whether it is frozen with PyInstaller 5.1 or 5.7 (and neither collects `api-ms-*` DLLs from basic python.org python environment).

Are you using python.org or anaconda python? Are you using a python package with binary extensions that require the `api-ms-*` DLLs?

**Reply by ghost** [reply]
I have always used python.org and frozen my applications with a clean virtual environment (venv). No external modules were installed.

I can confirm a basic ``print("hello world")`` program executes as intended, but not another program with the following imports for reference as it may be related to the problem. Do any of the following rely on the DLLs?

```py
import argparse
import ctypes
import os
import subprocess
```

**Reply by rokm** [reply]
```python
import argparse
import ctypes
import os
import subprocess
print("Hello!")
```
frozen under python.org 3.8.10 with PyInstaller 5.1 seems to run just fine in my test Win 7 VM.

`_ctypes` extension (used by `ctypes` module) depends on `libffi-7.dll`, but python 3.8 should bundle version that's compatible with Windows 7. The list of files collected in corresponding onedir build is:
``` 
base_library.zip
libcrypto-1_1.dll
libffi-7.dll
program.exe
python38.dll
select.pyd
unicodedata.pyd
VCRUNTIME140.dll
_bz2.pyd
_ctypes.pyd
_hashlib.pyd
_lzma.pyd
_socket.pyd
```
